File: sdk/android/frak-sdk/src/main/kotlin/id/frak/sdk/tracking/MergeSender.kt
package id.frak.sdk.tracking

import id.frak.sdk.core.FrakError
import id.frak.sdk.core.FrakLogger
import id.frak.sdk.identity.IdentityMerge
import id.frak.sdk.identity.ProofOp

/**
 * `POST /user/identity/merge/execute`. [QueuedRow.idempotencyKey] is the merge token itself; it
 * never reaches the wire from there — see [IdentityMerge.body].
 */
internal class MergeSender(
    private val logger: FrakLogger,
) : RowSender {
    // The backend mints a merge token with a 60-minute lifetime (AnonymousMergeService.ts:36); holding longer cannot succeed.
    override val holdTimeoutMillis: Long = 60L * 60 * 1000

    override suspend fun deliver(
        row: QueuedRow,
        ctx: SendContext,
    ): DeliveryOutcome {
        // A merge names the identity it folds in; a row that lost it can only post an empty
        // target, so drop it rather than spend the failure cap on a request that can't succeed.
        val anonymousId =
            row.clientId ?: run {
                logger.warn("Dropping a merge row that carries no anonymous id.")
                return DeliveryOutcome.Dropped
            }
        val mergeToken = row.idempotencyKey
        if (mergeToken.isEmpty()) {
            logger.warn("Dropping a merge row with no token.")
            return DeliveryOutcome.Dropped
        }
        val merchantId = row.merchantId ?: ctx.resolveMerchantId() ?: return DeliveryOutcome.Hold

        // Minted on every attempt, never cached: it binds merchantId, unknown until drain, and a
        // locked keystore is transient device state, not a verdict — Hold, not Rejected, or one
        // bad moment traps the row for good. Costs no round trip either way.
        val proof =
            ctx.signProof(ProofOp.Merge, merchantId, IdentityMerge.binding(mergeToken))
                ?: return DeliveryOutcome.Hold

        val body = IdentityMerge.body(mergeToken, anonymousId, merchantId, proof)
        val response =
            try {
                ctx.http.post(IdentityMerge.MERGE_EXECUTE_PATH, body.toString())
            } catch (failure: FrakError) {
                return DeliveryOutcome.Retryable(failure)
            }

        // A merge token is single-use, so 429/5xx must answer "later", not "no": that is the
        // difference between a retry and losing the referral. classifyStatus owns that boundary.
        val outcome = classifyStatus(response)
        if (outcome is DeliveryOutcome.Rejected) {
            logger.warn("Identity merge refused with status ${response.status}.")
        }
        return outcome
    }

    companion object {
        const val KIND = "merge"
    }
}

File: sdk/ios/Sources/FrakSDK/Tracking/MergeSender.swift
import Foundation

/// `POST /user/identity/merge/execute`. `QueuedRow.idempotencyKey` is the merge token itself; it
/// never reaches the wire from there — see `IdentityMerge.body`.
struct MergeSender: RowSender {
    static let kind = "merge"

    let logger: FrakLogger

    // The backend mints the token with exactly a 60-minute lifetime
    // (services/backend/src/domain/identity/services/AnonymousMergeService.ts:36); holding longer cannot succeed.
    var holdTimeout: TimeInterval { 60 * 60 }

    func deliver(row: QueuedRow, ctx: SendContext) async throws(CancellationError) -> DeliveryOutcome {
        // A merge names the identity it folds in; a row that lost it can only post an empty
        // target, so drop it rather than spend the failure cap on a request that can't succeed.
        guard let anonymousId = row.clientId else {
            logger.warn("Dropping a merge row that carries no anonymous id.")
            return .dropped
        }
        let mergeToken = row.idempotencyKey
        guard !mergeToken.isEmpty else {
            logger.warn("Dropping a merge row with no token.")
            return .dropped
        }

        let merchantId: String
        if let known = row.merchantId {
            merchantId = known
        } else if let resolved = await ctx.resolveMerchantId() {
            merchantId = resolved
        } else {
            return .hold
        }

        // Minted on every attempt, never cached: it binds merchantId, unknown until drain, and a
        // locked enclave is transient device state, not a backend verdict — hold, not rejected,
        // or one bad moment traps the row for good.
        guard let proof = await ctx.signProof(.merge, merchantId, IdentityMerge.binding(mergeToken)) else {
            return .hold
        }

        guard
            let body = IdentityMerge.body(
                mergeToken: mergeToken,
                anonymousId: anonymousId,
                merchantId: merchantId,
                proof: proof
            )
        else {
            return .dropped
        }

        do {
            let response = try await ctx.http.post(IdentityMerge.executePath, body: body)
            let outcome = classify(response)
            if case .rejected = outcome {
                logger.warn("Identity merge refused with status \(response.status).")
            }
            return outcome
        } catch let error as FrakError {
            return .retryable(error)
        } catch {
            // Cancellation, and nothing else — HTTPClient maps every transport failure to a FrakError.
            throw CancellationError()
        }
    }
}
